### PR TITLE
Add accurate pressure formulations

### PR DESCRIPTION
**Pull request by arvigj**

### CODE REVIEW

File: src/polyfem/assembler/PressureAssembler.hpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+		public:
+			ThermodynamicProcess() {}
+			virtual ~ThermodynamicProcess(){};
+			virtual double pressure(const double start_pressure, const double v0, const double v1) const { return 0; }
+			virtual double first_derivative(const double start_pressure, const double v0, const double v1) const { return 0; }
+			virtual double second_derivative(const double start_pressure, const double v0, const double v1) const { return 0; }
+			virtual double energy(const double start_pressure, const double v0, const double v1) const { return 0; }
+
+		protected:
+			const double atmospheric = 101.3e3;
```
Comment (teseoch) [inline src/polyfem/assembler/PressureAssembler.hpp:25]: this should be parameters?

File: src/polyfem/assembler/PressureAssembler.hpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+			virtual double second_derivative(const double start_pressure, const double v0, const double v1) const override
+			{
+				return gamma_ * (1 + gamma_) * (start_pressure + atmospheric) * std::pow(v0 / v1, gamma_) / v1 / v1;
+			}
+
+			virtual double energy(const double start_pressure, const double v0, const double v1) const override { return (start_pressure + atmospheric) * std::pow(v0, gamma_) * (std::pow(v1, 1. - gamma_)) / (1. - gamma_) - atmospheric * v1; }
+
+		private:
+			const double gamma_ = 1.4;
```
Comment (teseoch) [inline src/polyfem/assembler/PressureAssembler.hpp:70]: user defined?

File: src/polyfem/assembler/RhsAssembler.cpp [hunk 3]
```diff
@@ -671,6 +667,8 @@ namespace polyfem
 					for (int d = 0; d < size_; ++d)
 						rhs_fun.col(d) = rhs_fun.col(d).array() * weights.array();
 
+					const auto nodes = bs.local_nodes_for_primitive(primitive_global_id, mesh_);
+
```
Comment (teseoch) [inline src/polyfem/assembler/RhsAssembler.cpp:671]: why these changes?

File: src/polyfem/io/OutData.cpp [hunk 4]
```diff
@@ -94,7 +94,28 @@ namespace polyfem::io
 
 				for (int n = 0; n < normals.rows(); ++n)
 				{
-					normals.row(n) = normals.row(n) * vals.jac_it[n];
+					trafo = vals.jac_it[n].inverse();
```
Comment (teseoch) [inline src/polyfem/io/OutData.cpp:97]: was this not already fixed?

File: tests/test_diff.cpp [hunk 5]
```diff
@@ -559,80 +559,80 @@ TEST_CASE("shape-neumann-nodes", "[test_adjoint]")
...
 // }
 
-TEST_CASE("shape-pressure-neumann-nodes", "[test_adjoint]")
-{
-	const std::string path = POLYFEM_DATA_DIR + std::string("/differentiable/input/");
-	json in_args;
-	load_json(path + "shape-pressure-neumann-nodes.json", in_args);
-	auto state_ptr = create_state_and_solve(in_args);
-	State &state = *state_ptr;
+// TEST_CASE("shape-pressure-neumann-nodes", "[test_adjoint]")
```
Comment (teseoch) [inline tests/test_diff.cpp:562]: a lot of comments?

File: json-specs/input-spec.json [hunk 6]
```diff
@@ -1225,7 +1225,7 @@
     },
     {
         "pointer": "/contact/use_convergent_formulation",
-        "default": false,
+        "default": true,
```
[the remark this answers is not in the document]
Comment (arvigj) [reply json-specs/input-spec.json:1228]: Yes, I'll change this back.

File: src/polyfem/assembler/GenericProblem.cpp [hunk 7]
```diff
@@ -227,20 +233,46 @@ namespace polyfem
...
+						break;
+					}
+				}
+			}
+		}
+
+		double GenericTensorProblem::pressure_cavity_bc(const int boundary_id, const double t) const
+		{
+			Eigen::VectorXd pt;
+			pt.setZero(3);
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/assembler/GenericProblem.cpp:272]: Yea so this is for a closed cavity, the pressure does not depend on where in the cavity it is (constant inside the cavity).

File: src/polyfem/assembler/PressureAssembler.hpp (hunk 1 above)
Comment (arvigj) [reply src/polyfem/assembler/PressureAssembler.hpp:25]: I don't think so, this is pretty much constant at sea level on earth. We can rethink if there's very special circumstances to use something else.

File: src/polyfem/assembler/PressureAssembler.hpp (hunk 2 above)
Comment (arvigj) [reply src/polyfem/assembler/PressureAssembler.hpp:70]: Again, this is pretty much constant for normal air. Unless you plan on doing some very special things with specific single gases, I think this should remain hidden.

File: src/polyfem/io/OutData.cpp (hunk 4 above)
Comment (arvigj) [reply src/polyfem/io/OutData.cpp:97]: This was already fixed for the simulation, but I guess not fixed for the output.

File: tests/test_diff.cpp (hunk 5 above)
Comment (arvigj) [reply tests/test_diff.cpp:562]: Yea, so I'm waiting on this to get merged so I can send another PR with some optimization objectives that will increase the code coverage.

File: src/polyfem/State.hpp [hunk 8]
```diff
@@ -403,8 +414,14 @@ namespace polyfem
...
 		/// mapping from elements to nodes for neumann boundary conditions
 		std::vector<mesh::LocalBoundary> local_neumann_boundary;
+		/// mapping from elements to nodes for pressure boundary conditions
+		std::vector<mesh::LocalBoundary> local_pressure_boundary;
+		/// mapping from elements to nodes for pressure boundary conditions
+		std::unordered_map<int, std::vector<mesh::LocalBoundary>> local_pressure_cavity;
 		/// nodes on the boundary of polygonal elements, used for harmonic bases
 		std::map<int, basis::InterfaceData> poly_edge_to_data;
+		/// Matrices containing the input per node dirichelt
+		std::vector<Eigen::MatrixXd> input_dirichlet;
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/State.hpp]: Removed.

File: src/polyfem/State.hpp [hunk 9]
```diff
@@ -149,6 +150,8 @@ namespace polyfem
 		std::shared_ptr<assembler::MixedAssembler> mixed_assembler = nullptr;
 		std::shared_ptr<assembler::Assembler> pressure_assembler = nullptr;
 
+		std::shared_ptr<assembler::PressureAssembler> elasticity_pressure_assembler = nullptr;
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/State.hpp:153]: What do you mean? This needs to be stored to assemble the gradient and hessian at a later time.

File: src/polyfem/State.cpp [hunk 10]
```diff
@@ -1549,7 +1574,7 @@ namespace polyfem
 				solve_transient_navier_stokes(time_steps, t0, dt, sol, pressure);
 			else if (assembler->name() == "OperatorSplitting")
 				solve_transient_navier_stokes_split(time_steps, dt, sol, pressure);
-			else if (assembler->is_linear() && !is_contact_enabled()) // Collisions add nonlinearity to the problem
+			else if (assembler->is_linear() && !is_contact_enabled() && !is_pressure_enabled()) // Collisions add nonlinearity to the problem
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/State.cpp]: Done

**Review by Huangzizhou — APPROVED** [review]

File: src/polyfem/utils/ExpressionValue.cpp [hunk 11]
```diff
@@ -181,13 +227,44 @@ namespace polyfem
...
 		double ExpressionValue::operator()(double x, double y, double z, double t, int index) const
 		{
 			assert(unit_type_set_);
 
 			double result;
 			if (expr_.empty())
 			{
+				if (t_index_.size() > 0)
+				{
+					t = std::round(t * 1000.) / 1000.;
```
Comment (teseoch) [inline src/polyfem/utils/ExpressionValue.cpp:253]: what is this?

Comment (arvigj) [reply src/polyfem/utils/ExpressionValue.cpp:253]: This is the per timestep specification for boundary conditions. I think we agreed that it is brittle but we will refactor soon.

File: src/polyfem/State.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/State.hpp:153]: I don't understand what you mean by locally created?

File: src/polyfem/State.hpp [hunk 12]
```diff
@@ -403,6 +417,10 @@ namespace polyfem
 		std::vector<mesh::LocalBoundary> local_boundary;
 		/// mapping from elements to nodes for neumann boundary conditions
 		std::vector<mesh::LocalBoundary> local_neumann_boundary;
+		/// mapping from elements to nodes for pressure boundary conditions
+		std::vector<mesh::LocalBoundary> local_pressure_boundary;
+		/// mapping from elements to nodes for pressure boundary conditions
+		std::unordered_map<int, std::vector<mesh::LocalBoundary>> local_pressure_cavity;
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/State.hpp:423]: Yes, https://github.com/polyfem/polyfem/issues/326

File: src/polyfem/assembler/Problem.cpp [hunk 13]
```diff
@@ -22,22 +22,29 @@ namespace polyfem
 			std::vector<LocalBoundary> &local_boundary,
 			std::vector<int> &boundary_nodes,
 			std::vector<LocalBoundary> &local_neumann_boundary,
+			std::vector<LocalBoundary> &local_pressure_boundary,
+			std::unordered_map<int, std::vector<LocalBoundary>> &local_pressure_cavity,
```
[the remark this answers is not in the document]
Comment (arvigj) [reply src/polyfem/assembler/Problem.cpp:26]: Yes, https://github.com/polyfem/polyfem/issues/326
